lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,354 @@
+// 2. Filter from the temporary module operations that we are not interested in.
+// 3. Clone the temporary module into the context of the new container.
+//
+// Filtering in the source module and not after cloning into the new context
+// is important to avoid polluting the new context with types and attributes
+// not used by the remaining functions.
+std::unique_ptr<pipeline::ContainerBase>
+MLIRContainer::cloneFiltered(const pipeline::TargetsList &Filter) const {
+  auto DestinationContainer = std::make_unique<MLIRContainer>(name());
+
+  if (getModuleBlock(*Module).empty())
+    return DestinationContainer;
+
+  // The temporary module is automatically deleted at end of scope.
+  OwningModuleRef TemporaryModule(mlir::cast<ModuleOp>((*Module)->clone()));
+
+  bool RemovedSome = false;
+  visit(*TemporaryModule, [&](FunctionOpInterface F) {
+    if (F.isExternal())
+      return;
+
+    const MetaAddress MA = mlir::clift::getMetaAddress(F);
+    if (MA.isValid() and not Filter.contains(makeTarget(MA))) {
+      makeExternal(F);
+      RemovedSome = true;
+    }
+  });
+
+  if (RemovedSome)
+    pruneUnusedSymbols(*TemporaryModule);
+
+  MLIRContext &DestinationContext = *DestinationContainer->Context;
+  DestinationContext.appendDialectRegistry(Context->getDialectRegistry());
+
+  OwningModuleRef &DestinationModule = DestinationContainer->Module;
+  DestinationModule = cloneModuleInto(*TemporaryModule,
+                                      *DestinationContainer->Context);
+
+  return DestinationContainer;
+}
+
+void MLIRContainer::mergeBackImpl(MLIRContainer &&SourceContainer) {
+  if (getModuleBlock(*SourceContainer.Module).empty())
+    return;
+
+  if (getModuleBlock(*Module).empty()) {
+    Module = std::move(SourceContainer.Module);
+    Context = std::move(SourceContainer.Context);
+    return;
+  }
+
+  // Register the dialects of the other container in this container.
+  Context->appendDialectRegistry(SourceContainer.Context->getDialectRegistry());
+
+  // Clone the other container's module into this container's context.
+  // This module is automatically erased at the end of scope.
+  OwningModuleRef TemporaryModule = cloneModuleInto(*SourceContainer.Module,
+                                                    *Context);
+
+  mlir::Block &DestinationBlock = getModuleBlock(*Module);
+  visit(*TemporaryModule, [&](SymbolOpInterface Symbol) {
+    // Erase an existing symbol with the same name, if one exists.
+    if (auto S = SymbolTable::lookupSymbolIn(*Module, Symbol.getName()))
+      S->erase();
+
+    // Move each new symbol from the temporary module to the container's module.
+    Symbol->remove();
+    DestinationBlock.push_back(Symbol);
+  });
+
+  // Assume that at least some symbols were copied over and always prune.
+  pruneUnusedSymbols(*Module);
+}
+
+pipeline::TargetsList MLIRContainer::enumerate() const {
+  pipeline::TargetsList::List List;
+
+  visit(Module.get(), [&](FunctionOpInterface F) {
```

Comment:
Is this visit recursive? I'd be sad if we were to got through all the operation just to enumerate functions.

Reply:
No. There's a comment to that effect on `visit` itself.